### PR TITLE
Event generator (obsolete with CI4 Events)

### PR DESCRIPTION
**Pull request by objecttothis**
This is the CRUD event generator that @jekkos suggested be the driver for third-party integrations.  Per #2608 this has been implemented using the CI Events library and is pretty lean in its functionality so as not to bog the code down. The goal is to provide a layer of separation between the base code and third party integrations.

The concept is that "Triggers" are placed throughout the code (in the controllers) which trigger Create, Read, Update and Delete Methods.  Simultaneously "Listeners" receive the appropriate data and pass it off to the third-party integrations.

An example of this is that there is a trigger in the Items controller so that when a user successfully creates an item, the trigger puts the event out there with the item_data.  The event_create listener defined in the events/Integrations library then picks up the event and sends it to integrations_create defined in the same library which in turn will call any third party functions needing to use the data.  The idea is that that integrations_create function would do no heavy lifting but serve to call the third-party integrations functions only.  So, for example integrations_create($data) will look at $data[type] and when it's of type CUSTOMERS it might call, say mailchimp_add_email($data[data]) which would be defined separately.

There are a couple of question marks:
- Currently I have not added any "Read" triggers, because it was unclear to me how or if third-party integrations would use a read operation.  We can either remove the code or keep it as a place-holder for future use.  Alternately if someone can think of a good use case, I'll be glad to insert the triggers in the appropriate controllers.
- ~~Currently I have third-party integrations specific functions being created in the events/Integrations at the bottom of the class, but I'm thinking that will quickly get messy.  I'm thinking that we should extrapolate each integration into it's own file.  I was thinking that this could be done with a helper or library for each integration.  What is the best way to do this?  A helper would make the scope of each function global to the whole code, and I don't think we want that.  A library would require it be nested into its own class.  I'm not sure if that creates it's own scope problems.  Thoughts?~~ I went ahead and made changes so that each third party integration needs to be defined in its own library to create better code separation.
- This seems to be consistent with the rest of the code, but the errors on third-party integration fails are written to the log with a hard-coded English error message.  Should these be switched to toast notifications and not written in the log?  Another question is whether the error messages should be translatable regardless of whether they are toast notifications or written to the error log.  It seems that error log notifications are always in English and toast notifications are always translatable.

### CODE REVIEW

**Comment by objecttothis** [comment]
#2315, #2381 and #1680 are also relevant to the conversation.

File: application/controllers/Customers.php [hunk 1]
```diff
@@ -328,13 +339,21 @@ public function delete()
...
+				$event_failures = Events::Trigger('event_delete', array("type"=> "CUSTOMERS", "data" => $customers_to_delete), 'string');
+				
 				$count++;
 			}
 		}
-
+		
+		if($event_failures)
+		{
+		    log_message("ERROR","Third-Party Integration failed during Customer delete: $event_failures");
```
Comment (jekkos) [inline application/controllers/Customers.php]: I guess logging error messages in plain English is fine. Not sure but maybe there is a different method that directly logs the message as error (log_error or sth) so you can use one arg only?

Comment (jekkos) [reply application/controllers/Customers.php]: Wouldn't it make sense to handle the failure in the trigger call itself? I mean log the message there if a call to a specific listener fails?

It would mean less code duplication and a more detailed context ( the listener that failed the update ). Maybe you can extend the events library to do this logging or sth ( not sure how or it works exactly ).

File: application/controllers/Items.php [hunk 2]
```diff
@@ -432,62 +434,62 @@ public function attributes($item_id)
...
 				$link_value = $this->Attribute->get_link_value($item_id, $definition_id);
 				$values['selected_value'] = (empty($link_value)) ? '' : $link_value->attribute_id;
 			}
-
+			
 			if (!empty($definition_ids[$definition_id]))
 			{
 				$values['selected_value'] = $definition_ids[$definition_id];
 			}
-
```
Comment (jekkos) [inline application/controllers/Items.php]: Not sure if indentation is as it should be here? I think we use tabs but maybe double check it wasn't changed to spaces or sth.

File: application/controllers/Items.php [hunk 3]
```diff
@@ -613,34 +615,49 @@ public function save($item_id = -1)
...
+				}
+				else
+				{
+				    $event_failures = Events::Trigger('event_update', array("type"=> "ITEMS", "data" => $item_data), 'string');
+				}
+				
+				if($event_failures)
+				{
+				    log_message("ERROR","Third-Party Integration failed during item save: $event_failures");
+				}
```
Comment (jekkos) [inline application/controllers/Items.php]: Same remark as before, might not be the best place to log errors.

File: application/events/Integrations.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	 */
+	public function __construct()
+	{
+		//Third-Party Integrations by adding all code related to the integration into a library and referencing the library here with $this->ci->load->library();
+		$this->CI =& get_instance();
+		//$this->CI->load->library('LIB_NAME_HERE');
+		
+		Events::register('event_create', array($this,'integrations_create'));
+		Events::register('event_read', array($this,'integrations_read'));
+		Events::register('event_update', array($this,'integrations_update'));
```
Comment (jekkos) [inline application/events/Integrations.php]: Can't we make the events more specific? Like use items_update so that we don't need to do the huge switch a bit further on?

Or does that imply changing a lot of code?
Edit: the amount of events could explode so might not be the best idea.

File: application/events/Integrations.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	 */
+	public function integrations_update(array $data)
+	{
+		$results = NULL;
+		
+		//calls to update functions for the different Integrations
+		switch($data['type'])
+		{
+			case 'ITEMS':
+				break;
```
Comment (jekkos) [inline application/events/Integrations.php]: Maybe we can use a convention in naming the integration controller with the entity it will update and then lookup the file and call that method from here. Then we can cleanly keep related logic together in one file.

File: application/controllers/Customers.php (hunk 1 above)
Comment (objecttothis) [reply application/controllers/Customers.php]: > I guess logging error messages in plain English is fine. Not sure but maybe there is a different method that directly logs the message as error (log_error or sth) so you can use one arg only?

According to https://codeigniter.com/user_guide/general/errors.html there is just log_message.  There is show_error but that displays it to the screen.

Comment (objecttothis) [reply application/controllers/Customers.php]: > 
> 
> Wouldn't it make sense to handle the failure in the trigger call itself? I mean log the message there if a call to a specific listener fails?
> 
> It would mean less code duplication and a more detailed context ( the listener that failed the update ). Maybe you can extend the events library to do this logging or sth ( not sure how or it works exactly ).

The error message needs to be specific to where it failed, so if I refactor the error message inside the event (e.g. delete), then the error tells me information about the failure, but it doesn't tell me what kind of operation.  This is because multiple Controllers use the same events.  For example both Suppliers and Customers controllers have event_delete triggers, so if an error gets kicked back, it will only tell you that person so-and-so couldn't be deleted, but it won't tell you which controller called it.  Having the error where it is will kick a log message that tells you that it was the Customer Delete operation that person so-and-so couldn't be deleted.  But perhaps I didn't fully understand what you were suggesting, so let me know if you meant something else.

File: application/events/Integrations.php (hunk 4 above)
Comment (objecttothis) [reply application/events/Integrations.php]: > 
> 
> Can't we make the events more specific? Like use items_update so that we don't need to do the huge switch a bit further on?
> 
> Or does that imply changing a lot of code?
> Edit: the amount of events could explode so might not be the best idea.

Currently there are four different event types (one for each CRUD operation).  event_create, for example, is used by five different controllers, so refactoring the events out to more specific naming does make the number of events go up significantly and code re-use go down.  You wouldn't need the switch statements then, but unless I'm not looking at it right, you end up with more code instead of less.

> Maybe the switch can be made generic by dispatching to a method in the integrations controller the we define using a convention? Like if you want to hook items, you place the controller in a folder integration/Items.php and redefine the method you want to hook. That way you know to which method you can dispatch to from here.

Sorry, I don't completely understand.  Right now the switch has calls to each integration function based on the type.  For example, for the CLCdesq API I'm writing, in the 'ITEMS' switch I have $results = $this->CI->clcdesq_integration_lib->new_product_push($data['data']); which is defined in libraries/Clcdesq_integration_lib.php.  Can you clarify what you think should change?

File: application/events/Integrations.php (hunk 5 above)
Comment (objecttothis) [reply application/events/Integrations.php]: > 
> 
> Maybe we can use a convention in naming the integration controller with the entity it will update and then lookup the file and call that method from here. Then we can cleanly keep related logic together in one file.

Right now I have each integration in a library rather than a controller. The naming convention I'm using currently is [integration name]_integration_lib.php and as you've suggested, all member functions related to that integration need to remain in that class.  Do you think we should use a different naming convention?

Comment (daN4cat) [reply application/events/Integrations.php]: Not sure library is the correct format and location. This is an extension and libraries are more something you use in your code.

Comment (objecttothis) [reply application/events/Integrations.php]: > 
> 
> Not sure library is the correct format and location. This is an extension and libraries are more something you use in your code.

So, should it be a controller, as @jekkos referred to?

File: application/controllers/Items.php (hunk 2 above)
Comment (objecttothis) [reply application/controllers/Items.php]: > 
> 
> Not sure if indentation is as it should be here? I think we use tabs but maybe double check it wasn't changed to spaces or sth.

Not sure how those got inserted.  Fixed.

File: application/events/Integrations.php (hunk 4 above)
Comment (jekkos) [reply application/events/Integrations.php]: My idea was to organize the integration files bt grouping together the logic in files that have the same name as the file of the method it is triggered by.

Putting all logic in one library file will probably make the file too big.

You know which files you need to check for when you enter the switch statement.

File: application/controllers/Customers.php (hunk 1 above)
Comment (jekkos) [reply application/controllers/Customers.php]: Don't you already have all that context in the events lib? You have the switch case there which allows you to interpret the arguments using the correct context (whether it is a person, supplier)? You can reinterpret the argument accordingly.

It just feels cleaner as you will have a more generic way of handling the exception and the logic is where the call is dispatched.

Otherwise we need to add those if statements in every file separately.

File: application/events/Integrations.php (hunk 5 above)
Comment (jekkos) [reply application/events/Integrations.php]: It could be a controller or maybe just a class that extends a basic crud interface. I think we already have such an interface but not sure if it suits our need here.

Comment (daN4cat) [reply application/events/Integrations.php]: Extend by calling an interface. In the old days that was called "callback".

Comment (objecttothis) [reply application/events/Integrations.php]: > 
> 
> Extend by calling an interface. In the old days that was called "callback".

This is what I love about my job.  I get to learn stuff all the time.  I haven't touched interfaces since abstract classes with C++ in college, so I'll need to do some research and dive into it.  I admit that I have much to learn about software design, so I appreciate you and @jekkos helping with the design elements of this piece.  I'm glad to do the heavy lifting, but I might have some questions.  I don't see that we are using interfaces in other parts of the code here (at least my grep search for the interface keyword didn't return anything), so I'll need to look for example code elsewhere.  If you happen to have articles/examples you think translates well to what we are doing here, please feel free to send them along.

Comment (objecttothis) [reply application/events/Integrations.php]: BTW if anyone has a favorite software design text, article or otherwise, please feel free to send me the title/url.

Comment (daN4cat) [reply application/events/Integrations.php]: I'll lookup something over the weekend. This is the last week of work for the year so busy with social events and closing tasks.

Comment (jekkos) [reply application/events/Integrations.php]: Check the bottom of the `Secure_controller.php`. It contains the abstract definitions of save (write), view (read), search ( read multiple). I think we might just want to take that out again and move out to an interface in its own file. If the integrations 'controller' then implements this interface, we can find the right method to call from the events lib switch case. We might need to do a manual include and then cast to the interface before the call itself. Not sure if CI offers a better mechanism there, but this is already sth 'external' to their framework and we need to find the file anyway so I think this will be just fine.

Comment (jekkos) [reply application/events/Integrations.php]: Glad to help a bit on this one btw, if there's still stuh unclear I'll try to add some example code of what I have in mind for this `dispatch` thingy.

Comment (objecttothis) [reply application/events/Integrations.php]: > 
> 
> Glad to help a bit on this one btw, if there's still stuh unclear I'll try to add some example code of what I have in mind for this `dispatch` thingy.

@jekkos, thank you.  The thing that is most fuzzy for me right now is what the file/data structures of the integrations need to be.  Currently each integration is confined to its own library file in libraries/[INTEGRATION]_lib.php, but it sounds like perhaps each integration should have it's own Controller/Model pair?  If this is the case, should those be stored in the Controllers/Models directories respectively or should we come up with something more appropriate

The secondary part of this is the callbacks to determine what kind of CRUD operation it is (item, customer, etc) and tertiary would be error handling in a more elegant fashion to reduce code duplication.  In short, any of those areas that you want to help provide some more detail or code snippets, feel free.  I just learned about Github gists.  It might be easier to provide code snippets in a gist rather than in the comments here.

Comment (jekkos) [reply application/events/Integrations.php]: Well we could move this integration logic per enitity to some file that I'd call a 'controller hook'. I would like to write out a skeleton but as i don't have a decent keyboard at hand I fear this will be a bit hard.

We won't need new models as we will get all the data needed passed in through the call to the hook. What you will need to do is write some file lookup logic in the eventlib switch case that based on the case it is in

* Finds all integration controller hook files for that entity and calls `include($path)` 
* The $path can be built by looking for `integrations/*/$enitiy.php`. the wildcard is a globbing pattern and there is probably a way in php to find all files matching that pattern.
* Finds the method to call ( write, update, ...) in that loaded file (class) by casting it to the crud interface
* Calls the retrieved method within that
Casted object with the data that was passed from the original call in the controller ( could be an array with updated values )
* Catches any exception and log the message returned

File: application/controllers/Customers.php (hunk 1 above)
Comment (jekkos) [reply application/controllers/Customers.php]: @objecttothis still following? Or should I wrap up some code to show how this could work?

Comment (objecttothis) [reply application/controllers/Customers.php]: > @objecttothis still following? Or should I wrap up some code to show how this could work?

Sorry, I'm out of the office until January 6th. Feel free to post a gist of what an example would be and I'll do the grunt work from there.  Otherwise I'll try to figure it out when I get in on the 6th.

**Comment by jekkos** [comment]
I thought about this a bit further and think that the way we should try to make this code as modular as possible. When you add a 3rd party integration, you will need to sync with all data you already have in the ospos db. That's why we should use a publisher / subscriber pattern here. Currently I see following uses

* A publisher that reads existing records from the database and dispatches this data as an 'event' through the defined CRUD hooks. This one will allow you to sync existing data to an integration 'connector'. You'll actually replay all past events through this publisher by 
* Next we have the same events being fired from our controllers, meaning that once something is changed by a user in ospos, we can immediately dispatch those changes to the 3rd party integration.
* Also how should we hande an event dispatch failure? I guess that we will have to stop the sync at that point as our data could become inconsistent. All messages should be dispatched in order but the process should keep track of what was processed and ideally be able to resume from where it left of.

**Comment by jekkos** [comment]
Same idea for the mailchimp integration: it's failry limited at this time and will only add customers to mailchimp that are created in ospos, if you have the integration setup at that time. It seems there is no way of syncing data that is already there, which is a pity. So once we have the groundwork done here this should be reworked to use the plugin structure.

**Comment by daN4cat** [comment]
@jekkos yep, it's push only and not bidirectional syncing. The syncing requires comparisons and policies about what to do in conflicting scenarios.

**Comment by objecttothis** [comment]
Agreed. The event generator does not include an integration listener. So that you all know, this is high on my priority list, but I came down with a fever on Monday and am waiting for COVID-19 test results. I'm almost back to 100% and since I work from home, I will likely be back at it tomorrow. The bad news is that the crap has hit the fan and I am trying to get my staff geared to work from home.

**Comment by odiea** [comment]
Sorry to hear that. Get well soon.

File: application/controllers/Customers.php (hunk 1 above)
Comment (objecttothis) [reply application/controllers/Customers.php]: Changes made.  The error handling is now done in the trigger handler as suggested.

**Comment by objecttothis** [comment]
> 
> 
> I thought about this a bit further and think that we should try to make this code as modular as possible. When you add a 3rd party integration, you will need to sync with all data you already have in the ospos db. That's why we should use a publisher / subscriber pattern here. 
I apologize, I agree that the 3rd party integrations should be removed from the core code, which is why I wrote the code using event generators to trigger events.  The disadvantage of this method rather than a completely modular design is that this design requires the completion of 3rd-party integrations before the main application completes.  I can see long-running integrations, slowing the application down significantly.  A design that triggers an integration to run a playback of events would get around that, but it is out of my league to design and I would need some help with the code design.

> Currently I see following uses
> 
>     * A publisher that reads existing records from the database and dispatches this data as an 'event' through the defined CRUD hooks. This one will allow you to sync existing data to an integration 'connector'. You'll actually replay all past events through this publisher one-by-one.
I like the idea, but will need help with the design.

>     * Next we have the same events being fired from our controllers, meaning that once something is changed by a user in ospos, we can immediately dispatch those changes to the 3rd party integration.
> 
>     * Also how should we hande an event dispatch failure? I guess that we will have to stop the sync at that point as our data could become inconsistent. All messages should be dispatched in order but the process should keep track of what was processed and ideally be able to resume from where it left of.
The current method I am taking to handle integration failures is to report but not prevent the core application from processing the data.  The downside to this is that it potentially means stale data in the integration side.  We could potentially run the trigger as a transaction and fail the operation if the integration fails.  This would involve reworking the trigger location itself since I believe most or all are placed in the various controllers only after the application completes its work successfully, but not during its processing of the data.

Short form is that the biggest help I need is that we would reach a decision on exactly how the event generator should work, exactly how the design should be implemented (e.g., what goes in libraries, controllers, etc).

RFC: @jekkos @daN4cat

File: application/events/Integrations.php (hunk 4 above)
Comment (objecttothis) [reply application/events/Integrations.php:28]: > 
> 
> My idea was to organize the integration files by grouping together the logic in files that have the same controller name as the controller of the method it is triggered by.
> 
> So in this case we'd have sth like `integrations/clcdesq/Items.php` that contains all item crud related logic for that integration. You can then just include all Items.php in every subfolder of the `integrations/` folder and have the controller implement an interface that you call from the event lib.

@jekkos I think I see what you are getting at, but depending on the integration, it may be running the same or similar operations against multiple controllers.  For example, my CLCdesq integration on a Create operation (CRUD) builds a RESTful API request that utilizes operations against Items, Attributes, Suppliers and Config Controllers to gather all the data.   Right now the Class for my CLCdesq integration is sitting at 822 lines of code which is half of the mammoth Reports Controller.

> Putting all logic in one library file will probably make the file too big.
> 
> You know which files you need to check for when you enter the switch statement (Items.php, Suppliers.php, etc)

I don't know how much time you have on your hands, but I could use some help stubbing out what your dream is for the organization of this.  I want it to be well designed, and I get the sense you have more experience with it, but I also want to make sure I understand the implications of what you are proposing.

**Comment by jekkos** [comment]
I was checking whether it would not be better to  integrate an existing library to take care of this
example: https://eventsauce.io/

**Comment by jekkos** [comment]
Intereseting read: https://martinfowler.com/eaaDev/EventSourcing.html

**Comment by objecttothis** [comment]
> 
> 
> I was checking whether it would not be better to integrate an existing library to take care of this
> example: https://eventsauce.io/

The code in this PR is based off the existing library https://github.com/nathanmac/CodeIgniter-Events.  I'll take a look at Event Sauce.

**Comment by jekkos** [comment]
I think the eventsauce looks like it has some more functionality like support for testing, exception handling, event persistence, mesage generation from definition.. 
it feels like if we use something like this we don't need to reinvent the wheel as the current lib seems rather limited in that regard (just dispatching events but no clear way on how to do these things). Of course the specific integration with CI will remain (in this case wiring the 3rd party controllers as consumers) and handling event registration.

**Comment by objecttothis** [comment]
> 
> 
> I think the eventsauce looks like it has some more functionality like support for testing, exception handling, event persistence, mesage generation from definition..
> it feels like if we use something like this we don't need to reinvent the wheel as the current lib seems rather limited in that regard (just dispatching events but no clear way on how to do these things). Of course the specific integration with CI will remain (in this case wiring the 3rd party controllers as consumers) and handling event registration.

Agreed that EventSauce seems to be more robust.  A few questions come to my mind:
- What mechanism do 3rd party plugins need to use to run?  The current implementation triggers the event with the data passed to the appropriate integrations.  I skimmed the article that you sent and it appears that each 3rd party integration will need a parser to read the event log, but admittedly, I'm fuzzy on this, because it's new to me.
- Will the event logs give enough information for 3rd party integrations to do their work?  If the event log simply tells the integrations that an item was sold, without item_ids, sale amounts, etc, that could prove problematic.
- Are you available to help get this running?  I'm fine with getting my hands dirty, but I suspect this will go a lot slower if I need to implement all of it because of the learning curve.  
- The cost overhead in terms of memory and time for the current implementation is fairly low, because the data is being passed to the 3rd party integrations.  Does EventSauce add significant "go get it yourself" overhead against the database for the 3rd party integrations?

**Comment by jekkos** [comment]
Ok I think the next step would be to map out what we have in the PR here against what this library offers, as you suggest.

Basically this means redefining the events in this eventsauce DSL or yaml file and then running a code generator to see how it comes out. Once that is done we need to have some callbacks that push these events to the 3rd party API, and map them out to the correct fields on that side.

But I think that having event persistence could be an important one here, if it comes out of the box then that is quite a leap forward. As once an exception occurs, you know exactly up to which point you are synchronized, and can choose to resolve the error before you start a replay from that snapshot.
I'll see if I can free up some time maybe later next week to take a closer look. I can't make any hard promises there though, but I'll try my best.

**Comment by objecttothis** [comment]
> 
> 
> Ok I think the next step would be to map out what we have in the PR here against what this library offers, as you suggest.
> 
> Basically this means redefining the events in this eventsauce DSL or yaml file and then running a code generator to see how it comes out. Once that is done we need to have some callbacks that push these events to the 3rd party API, and map them out to the correct fields on that side.

sounds reasonable.  

> 
> But I think that having event persistence could be an important one here, if it comes out of the box then that is quite a leap forward. As once an exception occurs, you know exactly up to which point you are synchronized, and can choose to resolve the error before you start a replay from that snapshot.

Agreed, but it's unclear to me if each integration will have it's own log to draw on or if not, how one integration is supposed to keep track of what it has processed and what it hasn't.  e.g., integration A processes 10 of 20 events in the logfile does integration B still have those first 10 to process.  Also, does each integration need to have a listener?  Currently the integration is kicked off by the event generator, but event sourcing sounds more like eventsauce just writes the persistent log and the integrations are on their own.

> I'll see if I can free up some time maybe later next week to take a closer look. I can't make any hard promises there though, but I'll try my best.

Thanks.  My CLCdesq API 3rd party integration PR is running off this event manager, so merging it is dependent on this being in place.  I believe I understand the rough idea of Event sourcing, but my hesitation here is 25% not wanting to spend a bunch of time on another proof of concept if we don't know we are going to use it and 75% the learning curve.  Hence asking for your assistance with getting the big pieces in place and I can probably take it from there.  

If our bookstore didn't have the quarantine absolutely destroying our sales causing me to rush getting the API code written (CLCdesq API PR), I would have a go at it with just the understanding that this might just be learning a new tool, but I've got a crazy backlog of "urgent".

**Comment by objecttothis** [comment]
To clarify, clearly changing this component means I would adapt the CLCdesq API 3rd party integration.

**Comment by objecttothis** [comment]
> 
> 
> I see what you mean, you might want to have something minimal working first in that case so you can start using it already.

This PR as it stands now works.  I'm using it in #2754. 

> I understand the urgency and it's probably more important then to have a nicely designed framework for doing integrations.

I 100% agree that a nicely designed framework for integrations is absolutely important.  It sounds like EventSauce fits that description much better than the PR I have here.  We can take one of two routes.  

1. We can merge this PR as it is, and then rip it out and replace it when we get EventSauce in place.  The 3rd party integration code itself ideally wouldn't be affected anyway.
2. We can not merge this PR and I can do a rebase dance to the master, this PR and #2754 then use #2754 in production until we can get EventSauce running.

Thoughts?

> The integration you talk about is not something in this PR here I suppose (unless I missed that)? In that case I'd think it'd be just the mapping of these 'events' to the API that you are integrating?

#2754 is the 3rd party integration that I referred to.  I have just rebased that branch onto this one to make it work.
